kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeatures.java

```diff
@@ -414,250 +415,150 @@ public static Tuple2<Integer, Integer> minimumRequiredVersions(Set<TableFeature>
-   */
-  private static int getMinWriterVersion(String feature) {
-    switch (feature) {
-      case "inCommitTimestamp":
-        return 7;
-      default:
-        return 2;
-    }
+  public static boolean isRowTrackingEnabled(Protocol protocol) {
+    return protocol.getImplicitlyAndExplicitlyEnabledFeatures().contains(ROW_TRACKING_W_FEATURE);
+  }
+
+  public static boolean isDomainMetadataEnabled(Protocol protocol) {
+    return protocol.getImplicitlyAndExplicitlyEnabledFeatures().contains(DOMAIN_METADATA_W_FEATURE);
   }
 
   /**
-   * Determine whether a writer feature must be supported and enabled to satisfy the metadata
-   * requirements.
+   * Ensure all features listed in `currentFeatures` are also listed in `requiredFeatures`, or, if
+   * one is not listed, it must be capable to auto-update a protocol.
    *
-   * @param metadata the table metadata
-   * @param feature the writer feature to check
-   * @return whether the writer feature must be enabled
+   * <p>Note: Caller must make sure `requiredFeatures` is obtained from a min protocol that
+   * satisfies a table metadata.
    */
-  private static boolean metadataRequiresWriterFeatureToBeEnabled(
-      Metadata metadata, String feature) {
-    switch (feature) {
-      case "inCommitTimestamp":
-        return TableConfig.IN_COMMIT_TIMESTAMPS_ENABLED.fromMetadata(metadata);
-      default:
-        return false;
-    }
-  }
+  private void assertMetadataTableFeaturesAutomaticallySupported(
+      Set<TableFeature> currentFeatures, Set<TableFeature> requiredFeatures) {
+
+    Set<TableFeature> newFeatures = new HashSet<>(requiredFeatures);
+    newFeatures.removeAll(currentFeatures);
 
-  private static boolean isWriterFeatureSupported(Protocol protocol, String featureName) {
-    Set<String> writerFeatures = protocol.getWriterFeatures();
-    if (writerFeatures == null) {
-      return false;
+    List<TableFeature> nonAutoUpdateCapableFeatures =
+        newFeatures.stream()
+            .filter(f -> !(f instanceof FeatureAutoEnabledByMetadata))
+            .collect(toList());
+
+    if (!nonAutoUpdateCapableFeatures.isEmpty()) {
+      // TODO: fix this error message to be more informative, may be a Kernel exception
+      throw new UnsupportedOperationException(
+          "The current protocol does not support auto upgrading following features"
+              + nonAutoUpdateCapableFeatures);
     }
-    return writerFeatures.contains(featureName)
-        && protocol.getMinWriterVersion() >= TABLE_FEATURES_MIN_WRITER_VERSION;
   }
 
-  private static void validateNoInvariants(StructType tableSchema) {
-    if (hasInvariants(tableSchema)) {
-      throw DeltaErrors.columnInvariantsNotSupported();
+  /**
+   * Returns the smallest set of table features that contains `features` and that also contains all
+   * dependencies of all features in the returned set.
+   */
+  private static Set<TableFeature> getDependencyClosure(Set<TableFeature> features) {
```

Comment:
`getDependencyClosure` -- this doesn't mean much to me. is this a delta-spark convention? what does "closure" mean here?